### PR TITLE
Fix diff parsing when file contains loose carriage return

### LINKED ISSUE #1241: Index was outside the bounds of the array
GitHub extension version: 2.3.3.38
Visual studio version: 15.3.5

When I am opening a specific pull request from our GitHub Enterprise appliance I am getting this error in the UI.
 

image

```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at GitHub.Models.DiffUtilities.<ParseFragment>d__1.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.Exports\Models\DiffUtilities.cs:line 0
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitHub.InlineReviews.Services.DiffService.<Diff>d__2.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\DiffService.cs:line 31
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSessionService.<Diff>d__7.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSessionService.cs:line 51
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<UpdateFile>d__16.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 200
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<CreateFile>d__17.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 234
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<GetFile>d__9.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 86
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<GetFile>d__8.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 67
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.ViewModels.PullRequestDetailViewModel.<CreateChangedFilesTree>d__104.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.App\ViewModels\PullRequestDetailViewModel.cs:line 523
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at GitHub.ViewModels.PullRequestDetailViewModel.<Load>d__100.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.App\ViewModels\PullRequestDetailViewModel.cs:line 389    at GitHub.Models.DiffUtilities.<ParseFragment>d__1.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.Exports\Models\DiffUtilities.cs:line 0
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at GitHub.InlineReviews.Services.DiffService.<Diff>d__2.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\DiffService.cs:line 31
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSessionService.<Diff>d__7.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSessionService.cs:line 51
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<UpdateFile>d__16.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 200
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<CreateFile>d__17.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 234
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<GetFile>d__9.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 86
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.InlineReviews.Services.PullRequestSession.<GetFile>d__8.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.InlineReviews\Services\PullRequestSession.cs:line 67
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.ViewModels.PullRequestDetailViewModel.<CreateChangedFilesTree>d__104.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.App\ViewModels\PullRequestDetailViewModel.cs:line 523
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at GitHub.ViewModels.PullRequestDetailViewModel.<Load>d__100.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.App\ViewModels\PullRequestDetailViewModel.cs:line 389
```

Thanks!

### PR DESCRIPTION
**Pull request by jcansdale**
This is a fix for an issue where a diff contains one or more loose carriage returns.

Previous versions show the following when *any* file in the PR contained a loose loose carriage return. You can use the following PR as a repro: https://github.com/grokys/PullRequestSandbox/pull/39.

image

The above exception has been fixed as well as diff line matching. For example, take the following string: `"foo\r\r\nbar\n\nfo\ro\r\nbar\n"`.

When rendered in the GitHub diff view it looks like this:

image

However, when viewed using the GitHub file view it looks like this:

image

You can see that although there are only 5 numbered lines, the file actually takes up 7 lines (with two extra line breaks).

This is also how it appears when rendered on the Visual Studio diff (and normal) view. The only difference is that Visual Studio shows numbers for the rendered lines (as opposed to the line numbers when carriage return are ignored).

image

We need to account for this when the `OldLineNumber`  and `NewLineNumber` is calculated for a `DiffLine` (otherwise the lines that can be commented on won't match the lines displayed in Visual Studio).

This is a potential issue when lines are added to or deleted from a diff.

image

You can see on the two screen grabs above that the lines that can be commented on (the light colored ones) match the lines from the first screen grabs from GitHub's diff view. The line numbers are different, but the start of each line is consistent. This is how it's supposed to work in the fixed version.

NOTE: I wonder what other control chars we might be missing?

Fixes #1241

### CODE REVIEW

File: src/GitHub.Exports/Models/DiffUtilities.cs [hunk 1]
```diff
@@ -12,6 +12,16 @@ public static class DiffUtilities
 
         public static IEnumerable<DiffChunk> ParseFragment(string diff)
         {
+            // Turn Windows line endings into Unix line endings.
+            diff = NormalizeLineEndings(diff);
```
Comment (jcansdale) [inline src/GitHub.Exports/Models/DiffUtilities.cs]: I should create a `ContainsLooseCarriageReturn` regex/method before replacing all line endings.  This will be done for every file in PR. It's only required when there are actually loose carriage returns.

File: src/GitHub.Exports/Models/DiffUtilities.cs [hunk 2]
```diff
@@ -20,8 +30,14 @@ public static IEnumerable<DiffChunk> ParseFragment(string diff)
                 int oldLine = -1;
                 int newLine = -1;
 
+                // Diff lines should only be separated using a '\n' (lines can contain a '\r').
                 while ((line = reader.ReadLine()) != null)
                 {
+                    if (hasCarriageReturn)
+                    {
+                        line = UnescapeCarriageReturns(line);
```
Comment (jcansdale) [inline src/GitHub.Exports/Models/DiffUtilities.cs]: What if file already contained escaped carriage returns? Oops.

**Comment by jcansdale** [comment]
Files that contain a loose `\r` are very much the exception (there weren't any in any of our projects or repos I've looked at recently). The latest commit checks of any loose `\r` before making any accommodation for them. I was reluctant to implement a string reader for this one special and relatively unlikely case (the built in string reader is pretty optimized IIRC).

There is however an issue with the `\r` escaping and a custom string reader might be the cleanest fix. I'll take a look.

File: src/GitHub.Exports/Models/DiffUtilities.cs [hunk 3]
```diff
@@ -113,6 +135,37 @@ public static DiffLine Match(IEnumerable<DiffChunk> diff, IList<DiffLine> target
...
+        }
+
+        static string EscapeCarriageReturns(string text)
+        {
+            return text.Replace("\r", "\\r");
+        }
+
+        static string UnescapeCarriageReturns(string text)
+        {
+            return text.Replace("\\r", "\r");
```
Comment (grokys) [inline src/GitHub.Exports/Models/DiffUtilities.cs]: What will happen here if the diff already has the string "\\r" in it? It will get replaced by a control character, but will that affect the correctness of the parsing?

**Comment by shana** [comment]
> I was reluctant to implement a string reader for this one special and relatively unlikely case (the built in string reader is pretty optimized IIRC).

I don't think there's a need to implement a whole string reader tbh, you just want to go through the whole thing and stop at markers. You could even just do a `Split('\n')` and get all the lines like that because you know '\n' is always the last character of a line end regardless of whether it's windows or linux.

**Comment by jcansdale** [comment]
@shana, I've updated this according to your feedback. It's much cleaner now with no redundant string manipulation. It only chops out the sub-strings that it actually needs. 👍 

I did end up writing a simple line reader. There were a surprising number of edge cases in a small amount of code (that I wanted to test).

It only hunts explicitly for `\r` chars once and does so in an optimized way (using `IndexOf`).

File: src/GitHub.Exports/Models/DiffUtilities.cs (hunk 3 above)
Comment (jcansdale) [reply src/GitHub.Exports/Models/DiffUtilities.cs]: Yes, indeed. Poorly thought through escaping. This code is redundant now.

File: src/GitHub.Exports/Models/DiffUtilities.cs [hunk 4]
```diff
@@ -113,6 +114,65 @@ public static DiffLine Match(IEnumerable<DiffChunk> diff, IList<DiffLine> target
...
+
+                    index++;
+                }
+
+                return text.Substring(startIndex, length);
+            }
+
+            bool StartOfText => index == 0;
+
+            bool EndOfText => index == -1 || index == text.Length;
```
Comment (shana) [inline src/GitHub.Exports/Models/DiffUtilities.cs]: If null is passed in to the LineReader construction, this will throw.

File: test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs [hunk 5]
```diff
@@ -269,5 +315,33 @@ public void NoLineMatchesFromNoLines()
                 Assert.Equal(null, line);
             }
         }
+
+        public class TheLineReaderClass
+        {
+            [Theory]
+            [InlineData("", "", null, null)]
```
Comment (shana) [inline test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs]: Why not pass `new string[] { "", null }` directly into the test?

File: test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs [hunk 6]
```diff
@@ -269,5 +315,33 @@ public void NoLineMatchesFromNoLines()
...
+            [InlineData("1", "1", null, null)]
+            [InlineData("1\n2\n", "1", "2", null)]
+            [InlineData("1\n2", "1", "2", null)]
+            [InlineData("1\r\n2\n", "1", "2", null)]
+            [InlineData("1\r\n2", "1", "2", null)]
+            [InlineData("\r", "\r", null, null)]
+            [InlineData("\r\r", "\r\r", null, null)]
+            [InlineData("\r\r\n", "\r", null, null)]
+            [InlineData("\r_\n", "\r_", null, null)]
+            public void ReadLines(string text, string line1, string line2, string line3)
```
Comment (shana) [inline test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs]: There should also be a test checking what happens when null is passed into the constructor.

File: test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs (hunk 5 above)
Comment (jcansdale) [reply test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs]: I original wrote it using using NUnit, which allows test methods to accept `params string[] lines`. When I converted it to xUnit, I didn't realize you could define an array inside an attribute.

File: test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs [hunk 7]
```diff
@@ -269,5 +315,74 @@ public void NoLineMatchesFromNoLines()
...
+                Assert.Equal(expectCount, count);
+            }
+
+            [Fact]
+            public void CountCarriageReturns_NullText_ArgumentNullException()
+            {
+                Assert.Throws<ArgumentNullException>(() => DiffUtilities.LineReader.CountCarriageReturns(null));
+            }
+
+            static void Benchmark_ReadLineAndCountCarriageReturns_100000()
```
Comment (shana) [inline test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs]: Kill this method? It's not doing anything here.

File: src/GitHub.Exports/Models/DiffUtilities.cs [hunk 8]
```diff
@@ -113,6 +115,69 @@ public static DiffLine Match(IEnumerable<DiffChunk> diff, IList<DiffLine> target
             return null;
         }
 
+        public class LineReader
```
Comment (shana) [inline src/GitHub.Exports/Models/DiffUtilities.cs]: Since we ended up having a bunch of benchmarks for this, how about adding a comment here with a link to https://gist.github.com/shana/200e4719d4f571caab9dbf5921fa5276?

File: src/GitHub.Exports/Models/DiffUtilities.cs (hunk 4 above)
Comment (jcansdale) [reply src/GitHub.Exports/Models/DiffUtilities.cs]: Added null check to constructor.

File: test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs [hunk 9]
```diff
@@ -366,23 +366,6 @@ public void CountCarriageReturns_NullText_ArgumentNullException()
             {
                 Assert.Throws<ArgumentNullException>(() => DiffUtilities.LineReader.CountCarriageReturns(null));
             }
-
-            static void Benchmark_ReadLineAndCountCarriageReturns_100000()
-            {
-
-                var file = new System.Diagnostics.StackFrame(true).GetFileName();
```
Comment (jcansdale) [inline test/GitHub.InlineReviews.UnitTests/Models/DiffUtilitiesTests.cs]: This was an ad-hoc test method I ran using TestDriven.Net. Easy for quick benchmarking. Zapped it now.
